Code review

dpsmith opened the pull request:
These are changes requested during code review of Linux kernel.

dpsmith committed:
crb: fixing signedness of cmd_ready

Signed-off-by: Daniel P. Smith <dpsmith@apertussolutions.com>

dpsmith committed:
tpm: simplify pcr_extend function

This reduces the multiple assignments of the algorithm down to one.

Signed-off-by: Daniel P. Smith <dpsmith@apertussolutions.com>

dpsmith committed:
tpm2_constants: adopt the linux _AT() macro

Signed-off-by: Daniel P. Smith <dpsmith@apertussolutions.com>

dpsmith committed:
core: refactored hw interface

This commit refactors how the library interfaces with the hardware
interfaces. In particular it abstracts the interfaces into an
operations callback interface on struct tpm. This eliminates a lot
of the switches on tpm->intf to select the correct function for the
hardware interface. Additionally this refactoring enabled cleaning
up some areas of the code logic.

Signed-off-by: Daniel P. Smith <dpsmith@apertussolutions.com>

rossphilipson commented on crb.c:
```diff
@@ -303,3 +279,32 @@ size_t crb_recv(struct tpmbuff *buf)
+	u8 i;
+	struct tpm_crb_intf_id_ext id;
+
+	for (i = 0; i <= TPM_MAX_LOCALITY; i++)
```
I had to remove this from tis_init(). Trying to close some localities on some systems causes crashes. See:
a0ba8b6b999493e85b6d098dadadc3d5a419e060

dpsmith replied:
ack

rossphilipson commented on tis.c:
```diff
@@ -191,33 +177,50 @@ size_t tis_recv(enum tpm_family f, struct tpmbuff *buf)
+		return 0;
+
+	t.ops.request_locality = tis_request_locality;
+	t.ops.relinquish_locality = tis.relinquish_locality;
```
That "." looks like a type-o

dpsmith replied:
ack

rossphilipson replied:
Still needs fixing?

dpsmith replied:
Yep along with other silliness which is what I am working to get cleaned up.

rossphilipson commented on tpm.c:
```diff
@@ -58,51 +58,29 @@ struct tpm *enable_tpm(void)
 	find_interface_and_family(t);
 
 	switch (t->intf) {
-	case TPM_DEVNODE:
```
I thought you were going to leave DEVNODE and UEFI in tpmlib and strip them from the kernel versions?

dpsmith replied:
Nope, stripping them completely for now. Will bring them back along with a working implementation instead of the stubs.

rossphilipson commented on tpm.c:
```diff
@@ -137,16 +102,14 @@ int tpm_extend_pcr(struct tpm *t, u32 pcr, u16 algo,
 	int ret = 0;
 
 	if (t->buff == NULL) {
```
Ditch the brackets for single line in if statement.

dpsmith replied:
ack

rossphilipson commented on tpm.c:
```diff
@@ -137,16 +102,14 @@ int tpm_extend_pcr(struct tpm *t, u32 pcr, u16 algo,
 		if (algo != TPM_ALG_SHA1) {
-			ret = -EINVAL;
-			goto out;
+			return -EINVAL;
```
Ditto on brackets anywhere the if just became a one line return statement.

dpsmith replied:
ack

rossphilipson commented on tpm.c:
```diff
@@ -160,37 +123,32 @@ int tpm_extend_pcr(struct tpm *t, u32 pcr, u16 algo,
 
 		ret = tpm2_extend_pcr(t, pcr, d);
 	} else {
 		ret = -EINVAL;
```
This can be an inline return and then you can ditch the variable ret.

dpsmith replied:
but ret is set in tpm1.2 and at the end of tpm2, i would have to make all of those inline returns as well which I am not sure that will be in line with kernel code style.

rossphilipson replied:
Oh I missed that. Never mind then.

rossphilipson commented on tpm1_cmds.c:
```diff
@@ -64,69 +64,29 @@ int tpm1_pcr_extend(struct tpm *t, struct tpm_digest *d)
-		ret = -ENOMEM;
-		goto out;
-	}
+	tpmb_trim(tpmb_size(b));
```
I guess trim just resets the whole buffer if you trim the entire thing?

dpsmith replied:
The trim method shrinks the buffer usage, so this should trim back to the beginning if I didn't make a mistake elsewhere in the put method. A lot shorter than tpmb_free line followed by tpmb_alloc line.

rossphilipson commented on tpm1_cmds.c:
```diff
@@ -64,69 +64,29 @@ int tpm1_pcr_extend(struct tpm *t, struct tpm_digest *d)
+	tpmb_trim(tpmb_size(b));
 
 	hdr = (struct tpm_header *)b->head;
+	tpmb_put(b, sizeof(struct tpm_header));
```
Was this just missing from the original code and just happened to work because the stale values were hanging around?

dpsmith replied:
No, that put was hidden in the tpmb_alloc call that would have been called in the previous version.

rossphilipson commented on tpm2_auth.c:
```diff
@@ -40,7 +40,7 @@ u8 *tpm2_null_auth(struct tpmbuff *b)
 	u32 *handle;
 	u8 *auth = (u8 *)tpmb_put(b, NULL_AUTH_SIZE);
 	if (auth == NULL)
```
You will probably get nit picks on annoying things like "this should be if (!auth)". Also just return NULL here.

dpsmith replied:
ack

rossphilipson commented on tpm2_cmds.c:
```diff
@@ -35,6 +35,9 @@
 		u32 code)
 {
+	/* ensure buffer is free for use */
+	tpmb_free(b);
```
Should this be done in the tpm1_cmd.c code too?

dpsmith replied:
Actually, this along with the next couple of lines are a better pattern. Will use it in tpm1_cmd.c

rossphilipson commented on tpm.h:
```diff
@@ -25,23 +25,29 @@
+	u8 request_locality(u8);
+	void relinquish_locality(void);
+	size_t send(struct tpmbuff *);
+	size_t recv(enum tpm_family, struct tpmbuff *);
```
crb_recv() does not match this function pointer signature. You need to add family. I would also add __attribute__((unused)) to the two params to make sure there are no warnings,

dpsmith replied:
ack

rossphilipson commented on tpm2_constants.h:
```diff
@@ -9,57 +9,62 @@
  */
 
+
+#ifndef _AT
```
Why is this before #ifndef _TPM2_CONSTANTS_H?

dpsmith replied:
I was being sloppy ( ._.)

dpsmith committed:
core: a few minor code clean up changes

This commit contains a few minor code clean up changes that either
improve stablity or code hygine.

Signed-off-by: Daniel P. Smith <dpsmith@apertussolutions.com>

rossphilipson commented on crb.c:
```diff
@@ -298,8 +274,34 @@ size_t crb_send(struct tpmbuff *buf)
 }
 
-size_t crb_recv(struct tpmbuff *buf)
+size_t crb_recv(__attribute__((unused)) enum tpm_family, __attribute__((unused)) struct tpmbuff *buf)
```
Maybe bring the second arg onto its own line so this line is not so long...

rossphilipson commented on crb.c:
```diff
@@ -298,8 +274,34 @@ size_t crb_send(struct tpmbuff *buf)
-size_t crb_recv(struct tpmbuff *buf)
+size_t crb_recv(__attribute__((unused)) enum tpm_family, __attribute__((unused)) struct tpmbuff *buf)
 {
 	/* noop, currently send waits until execution is complete*/
```
Not sure what this comment means. Is there no such thing as a CRB receive operation?

dpsmith replied:
Not really, you just read the CRB buffer back and specifically in the case of an extend the return is less than the CRB buffer so its more about just waiting for the TPM state to change.

rossphilipson commented:
I have looked this over carefully. It still has a few minor issues but I am going to merge it and then fix those (and PR them).

Reviewed-by: Ross Philipson <ross.philipson@oracle.com>
